taichi/transforms/loop_invariant_code_motion.cpp

```diff
@@ -0,0 +1,137 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/system/profiler.h"
+
+#include <stack>
+
+TLANG_NAMESPACE_BEGIN
+
+class LoopInvariantCodeMotion : public BasicStmtVisitor {
+ public:
+  std::stack<Block *> loop_blocks;
+
+  const CompileConfig &config;
+
+  DelayedIRModifier modifier;
+
+  LoopInvariantCodeMotion(const CompileConfig &config) : config(config) {
+    allow_undefined_visitor = true;
+  }
+
+  bool stmt_can_be_moved(Stmt *stmt) {
+    if (loop_blocks.size() <= 1 || (!config.move_loop_invariant_outside_if &&
+                                    stmt->parent != loop_blocks.top()))
+      return false;
+
+    bool can_be_moved = true;
+
+    Block *current_scope = stmt->parent;
+
+    for (Stmt *operand : stmt->get_operands()) {
+      if (operand->parent == current_scope) {
+        // Has an input that is in the current scope
+        // Operand can either be in the current scope or a outer scope
```

Comment:
```suggestion
        // This statement has an operand that is in the current scope,
        // so it can not be moved out of the scope.
```

I just thought about a case:
```python
for i in range(10):
  a = x[0]
  if b:
    c = a + x[1]  # a is not an operand in the current scope, but we cannot move it out.
```
(`x` is a global field preventing other optimizations such as constant fold)

Would you please add some tests like this? I'm not sure if we need something like
https://github.com/taichi-dev/taichi/blob/a4279bd8e1aa6b7d0d269ff909773d333fab5daa/taichi/analysis/verify.cpp#L42
to check if the operand is visible outside the loop.

Reply:
For this case I thought about it but I think `x[1]` should be an global load operation right? Since we don't move load store operations, then `c = a + x[1]` can not be moved as `x[1]` is in the `if` scope.

Reply:
You’re right, but what about when the statement is `c = a + 1`, and the constant 1 is outside the if clause?

Reply:
I don't quite know why the statement can't be moved if that's the case?

Reply:
```python
for i in range(10):
  a = x[0]
  if b:
    c = a + 1
    print(c)
```
If we move `c = a + 1` out of the loop:
```python
c = a + 1  # a is not visible here!
for i in range(10):
  a = x[0]
  if b:
    print(c)
```

Reply:
Oh yeah you are right, this does need a check. This is only for statements within a if block nested in a loop block right?

Reply:
Right, we need to flag loads from read-only data and move those as well. I think we can add it in another PR.